### DISCUSSION TITLE
Making unit tests mandatory for write-related PR

### DISCUSSION THREAD

**Discussion by nkosi23** [seaweedfs/seaweedfs#6161] [General]
Hi there! 

I have noticed that there has been a couple of data loss bug in the recent past mainly due to regressions. And i was surprised that PRs introducing bugs / regressions in core features like indexing were able to be merged so easily.

So i thought it'd be a good idea to start this thread regarding how stability could be improved moving forward, and also to clarify what actually needs to be improved in terms of process, and what is just a matter of perception.

For starters I have the following remarks/questions:

- what about introducing a mandatory unit test policy for PRs impacting write-related features, mainly to ensure that no-regression is introduced (which  in my opinion, for a storage system is almost more important than testing the new feature)? At least for the core features to begin (barebone master + volume server management without filer - and maybe also cover EC)
- does seaweedfs have a suitable/satisfying test harness in your opinion? (Including when it comes to load testing, testing large amounts of data, testing backward compatibility etc...). Otherwise, the community could try to find solutions in that direction, from identifying cloud providers able to provide the CI infrastructure needed by the project for free, to contributing test code.

I strongly feel that seaweedfs would greatly benefit from emphasing stability, at least for core features. This is quite an old project so by now there should be a happy path that users feel 2000% confident about. 

Because while this project is great, it is also at times scary to be honest. I always spend a couple of weeks/months studying bug reports before committing to using something in production. And i often feel that i will not have the brain required to fix the situation if something goes wrong with seaweedfs especially since things like an easy way to backup / restore are not documented, and even using core features (no filer, no EC etc...) does not seem to guarantee that catastrophic data loss regressions won't hit me.

When i use other storage systems such as postgresql or even more recent ones like couchdb, i never worry about data loss due to regressions between new versions, so my goal with this thread is to modestly contribute to getting seaweedfs to this place, if it isn't already.

This project is feature-rich and outstanding, but regressions impacting data safety can give the perception that it does not have the level of maturity / reliability it should have after all these years, which is unfortunate.

**Comment by chrislusf** [comment]
Let's have actionable items. For example, which bug could have been avoided by a unit/integration test. And send PRs to add the tests.

**Reply by nkosi23** [reply]
This one for example: https://github.com/seaweedfs/seaweedfs/issues/5215

But with respect, my goal here is not to have a blame game or to send PRs after the fact, but to have a discussion to know the opinions of people - first of all active contributors - on this question.

I just want to have a friendly discussion going on this topic. I am involved with a number of open-source projects on Github and it is very common that maintainers are aware of this kind problems, and have a wish-list that they wished they could go through if they had more time/contributors. The goal here is to gather these high-impact needs in the area of stability / testing, to then turn them into actionable items. 

I gather from the tone of your answer that for you - the author and lead maintainer - the situation is fine and this kind of policy is not needed? If so I have no problem with that. I was also specifically asking, if you feel help would be needed to improve the test harness or if it is already fine? 

People - may think that I am overstating the problem, and telling me so is also perfectly fine. The most important thing is that we are able to discuss this and clarify the situation in a friendly manner. Before coding, we need to look at the big picture as well. For starters you likely have an opinion on what needs to be done, or items you never had the time to get at.

Also, I may not have the technical ability to send PRs for everything, but if needs in this area are clarified, more people will contribute, since it is in everyone's interest to make things as stable as possible.

**Reply by nkosi23** [reply]
Also in terms of PR, since you are very familiar with the code, could you list areas where you think better test coverage would be needed? We can also approach the discussion from this angle if you prefer.

**Reply by kmlebedev** [reply]
Hello and thank you for this discussion, I also encountered regression a couple of years ago and then focused on adding unit s3 tests. But as we have seen now, this is not enough, since they do not use test data from the previous version.

**Reply by kmlebedev** [reply]
now yes there are not enough tests for background operations such as vaccum, weed fix, generate EC volume and this needs to be done for idx with storage type in memory and in leveldb. I have problems with integrity of dat files at the moment when the calculated size according to idx does not correspond to the size of the data in the dat file. I would like to have a separate tool for diagnostics and correction of dat file with corresponding tests for 30Gb and 800Gb versions.

**Reply by chrislusf** [reply]
For example, https://github.com/seaweedfs/seaweedfs/discussions/6162 needs an integration test to verify the bug and fix it.

**Reply by evgmik** [reply]
I think everyone would agree that a full test suite would be fantastic. But writing such is even more work than the main code plus it requires a full understanding of internals. I am sure we are experienced enough to know that having a test next to the fix is a good idea. But this is very challenging. I was able to fix a few bugs (and hopefully did not introduce new ones :) But I was not able to write a test, since it involved a lot of moving parts.

I would personally love to have more tests about potential bitrot detection. But @chrislusf  already doing heroic amount of work, so I am not going to ask for anything which I am not willing to do myself.
